extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/TransactionEntityManagers.java

```diff
@@ -32,9 +32,18 @@ public TransactionEntityManagers() {
     }
 
     public EntityManager getEntityManager(String unitName) {
-        return managers.computeIfAbsent(unitName,
-                un -> new TransactionScopedEntityManager(tm, tsr, jpaConfig.getEntityManagerFactory(un), unitName,
-                        requestScopedEntityManagers));
+        TransactionScopedEntityManager ret = managers.get(unitName);
```

Comment:
This should just be changed to be a concurrent hash map, and TransactionEntityManagers should really be explicitly marked as being ApplicationScoped to make it clear how the lifecycle works.

Reply:
From what I saw (i.e. 3 `ConcurrentModificationException` being thrown in a very short amount of time) I concluded, that this is a very often invoked code-path. But the bug only affects the first few invocations, after that it should be fine (I guess).

I did NOT benchmark this solution against a `ConcurrentHashMap` one, but I can if needed. My strain of thought was to just avoid `ConcurrentModificationException` without affecting performance to much.  Therefore I went for the least intrusive way (i.e. a double-checked write lock) to avoid a major performance hit, as I guessed a `null` check is quicker then a fully fledged `ConcurrentHashMap`.

Reply:
In Hibernate Validator, we switched to the following pattern for this sort of things:
```
private ConcurrentHashMap map...;
```

And then:
```
Object entry = map.get("");
if (map != null) {
    return entry;
}
return map.computeIfAbsent(...);
```

From the benchmarks we run, it's faster than using `computeIfAbsent()` directly (at least with JDK 8).

Reply:
@gsmet That seems like a decent approach.

The thing I am most skeptic about is, if `ConcurrentHashMap.get()` is measurably slower then `HashMap.get()`, which would mean that performance is slower on every call, not just on the first few.

I will try to find some time to run some benchmarks against all 3 approaches (plain `HashMap`, my pull-request, `ConcurrentHashMap` with @gsmet extension). I'll let you know what I find out.

Reply:
@j-be would you mind implementing the approach I presented and run the benchmark afterwards?

The code is obviously wrong so I think we should fix it for 0.17.0 with a decent approach. Then we can run benchmarks and tune the approach.

0.17.0 is planned for Wednesday so we would need a patch for tomorrow. If you can't make it, I'll do it.

Thanks!

Reply:
Sure thing. Implemented as @gsmet proposed. First (nothing more then a smoke-test) benchmarks show no difference between the `ConcurrentHashMap` implementation and my initial one (Requests per second: 8784.85 vs. 8755.22 is well within the margin of error).